### PR TITLE
built-in retries for failed fetches

### PR DESCRIPTION
**Pull request by williamboxhall**
This doesn't even compile and it's very ugly but I ran out of time to keep working on it today so I thought I'd throw it to you for comment/to keep working on it if you want to.

You should be able to get the idea based off the new method I've added to ClumpSource.

@fwbrasil let me know what you think?

### CODE REVIEW

File: src/main/scala/clump/ClumpFetcher.scala [hunk 1]
```diff
@@ -35,6 +39,20 @@ final class ClumpFetcher[T, U](source: ClumpSource[T, U]) {
     results
   }
 
+  private def retryIfNeccessary(maybeFailed: Future[Map[T, U]], retries: Int) = {
```
Comment (williamboxhall) [inline src/main/scala/clump/ClumpFetcher.scala]: I didn't have time to implement this recursion properly obviously :P but you can probably see what I am trying to do

**Comment by fwbrasil** [comment]
Nice feature! Two suggestions:

1. Use a partial function to define the ```maxRetries``` per exception:

```scala
source.maxRetries {
   case e: SomeException => 10
}
````

2. Add the retry logic to ```fetchBatch```, not ```flushInBatches```.  We don't need to retry all fetches, only the individual fetch.

I'll try to work on this today, but I'm not sure I'll have time.

**Comment by fwbrasil** [comment]
Another alternative for the ```maxRetries``` interface:

```scala
source.maxRetries[SomeException](10)
```

**Comment by williamboxhall** [comment]
I like both interfaces, the partial function gives you flexibility in what
to match, the second is more terse. Maybe you could have both? Go ahead and
implement if you have time

By the way, I think we can drop the "max" on both of these and have
.batchSize and .retries

**Comment by fwbrasil** [comment]
@williamboxhall It compiles now but there aren't tests yet.

**Comment by williamboxhall** [comment]
Wow awesome @fwbrasil, really elegant implementation. I would love to add some tests for you but I suspect I will not have much time over the next few days before I leave for Berlin

**Pull request merged**